### PR TITLE
[MPC] crs generation

### PR DESCRIPTION
**Pull request by dtebbs**
Load output from powersoftau and compute a CRS for a QAP by performing the appropriate linear combination and simulating the second layer of the MPC.
Unoptimized at this stage.

### CODE REVIEW

**Comment by AntoineRondelet** [comment]
Depends on #40

**Comment by AntoineRondelet** [comment]
Since this pull request is now specific to the MPC, I think it makes sense to:
- Update the mpc branch to the the latest commit on develop
- Change the source branch of this PR to be `mpc` so that we can "isolate" the mpc specific pull requests and have them all merged to develop when it makes sense.

What do you think @dtebbs ?

**Comment by dtebbs** [comment]
> Since this pull request is now specific to the MPC, I think it makes sense to:
> ...
> What do you think @dtebbs ?

Yes, absolutely.  Updated as you describe.

File: src/snarks/groth16/powersoftau_utils.cpp [hunk 1]
```diff
@@ -0,0 +1,355 @@
...
+
+namespace libzeth
+{
+
+namespace
+{
+
+// Utility functions
+
+struct membuf : std::streambuf {
```
Comment (AntoineRondelet) [inline src/snarks/groth16/powersoftau_utils.cpp]: Let's stick to `class` with public members

File: src/snarks/groth16/multi_exp.hpp [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+///
+template<typename ppT, typename GroupT>
+extern GroupT multi_exp(
+    const std::vector<GroupT> &gs, const libff::Fr_vector<ppT> &fs)
+{
+    assert(gs.size() >= fs.size());
+    assert(gs.size() > 0);
+
+#if 0
```
Comment (AntoineRondelet) [inline src/snarks/groth16/multi_exp.hpp]: What was your intention here? This block never gets executed (not even compiled). Maybe we can introduce a compilation variable if you want to add the binary of this fragment to the binary.

File: src/snarks/groth16/multi_exp.hpp [hunk 3]
```diff
@@ -0,0 +1,44 @@
+#pragma once
```
Comment (AntoineRondelet) [inline src/snarks/groth16/multi_exp.hpp]: Why use external linkage and use `pragma once` instead of having these functions defined in a cpp file and include the header where needed?

File: src/snarks/groth16/evaluation_from_lagrange.tcc [hunk 4]
```diff
@@ -0,0 +1,37 @@
+#pragma once
```
Comment (AntoineRondelet) [inline src/snarks/groth16/evaluation_from_lagrange.tcc]: I'd stick to:
```cpp
#ifndef __FILE_EXTENSION__
#define __FILE_EXTENSION__ 
...
#endif
```

This is applicable to several files in this PR that also use `#pragma once` as an alternative

File: pyClient/testZethUtils.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
+import zethUtils
```
Comment (AntoineRondelet) [inline pyClient/testZethUtils.py]: All these python files are not related to this PR. I guess you rebased this one on top of develop and haven't rebased mpc. Would you mind updating mpc so we do not introduce these changes a second time? (it would remove non-related files from the review also)

File: src/snarks/groth16/evaluation_from_lagrange.tcc (hunk 4 above)
Comment (AntoineRondelet) [reply src/snarks/groth16/evaluation_from_lagrange.tcc]: Then I'd include the `.tcc` file at the end of the `.hpp` file

File: src/snarks/groth16/mpc_utils.cpp [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+using ppT = libff::default_ec_pp;
+using Fr = libff::Fr<ppT>;
+using G1 = libff::G1<ppT>;
+using G2 = libff::G2<ppT>;
+
+// -----------------------------------------------------------------------------
+// mpc_layer1
+// -----------------------------------------------------------------------------
+
+mpc_layer1::mpc_layer1(
```
Comment (AntoineRondelet) [inline src/snarks/groth16/mpc_utils.cpp]: With `mpc_layer1` are you referring to the `Round 1` of the mpc (namely "Powers of Tau", 6.1 in the paper) + The result of the computation of the linear combinations between the rounds (6.2 in the paper)?

File: src/snarks/groth16/mpc_utils.cpp [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+
+        ABC_i_g1[i] = beta_A_at_t + alpha_B_at_t + C_at_t;
+    }
+
+    assert(num_variables + 1 == A_i_g1.size());
+    assert(num_variables + 1 == B_i_g1.size());
+    assert(num_variables + 1 == B_i_g2.size());
+    assert(num_variables + 1 == ABC_i_g1.size());
+
+    // TODO: Sparse B
```
Comment (AntoineRondelet) [inline src/snarks/groth16/mpc_utils.cpp]: What do you mean?

File: src/snarks/groth16/mpc_utils.hpp [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+
+#include <vector>
+
+// Structures and utility functions related to CRS generation via an
+// MPC.
+
+namespace libzeth
+{
+
+using ppT = libff::default_ec_pp;
```
Comment (AntoineRondelet) [inline src/snarks/groth16/mpc_utils.hpp]: The templates are "instantiated" in the `prover_server.cc` in order to make sure it is consistent everywhere. I can see that you have re-instantiated your templates in several files here. I'd move all these `using` directives to the program's entrypoint to ensure a consistent type handling (and I find it easier to reason about). We might want to keep the same template name and change the type with which we instantiate it for some exceptional cases, but here it does not seem to be one.

File: src/snarks/groth16/mpc_utils.cpp [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+#include "multi_exp.hpp"
+
+#include <libff/algebra/scalar_multiplication/multiexp.hpp>
+
+using namespace libsnark;
+
+namespace libzeth
+{
+
+using ppT = libff::default_ec_pp;
```
Comment (AntoineRondelet) [inline src/snarks/groth16/mpc_utils.cpp]: Remove these `using` directives. They are already specified in the entrypoint (see: `src/test/mpc_test.cpp`). Then, rename file to `.tcc` file and "templatize" the functions and classes.

File: src/snarks/groth16/mpc_utils.cpp [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+    // Langrange polynomials, and therefore A, B, C will have order
+    // (n-1).  T has order n.  H.t() has order 2n-2, => H(.) has
+    // order:
+    //
+    //   2n-2 - n = n-2
+    //
+    // Therefore { t(x) . x^i } has 0 .. n-2 (n-1 of them), requiring
+    // requires powers of tau 0 ..  2.n-2 (2n-1 of them).  We should
+    // have at least this many, by definition.
+
```
[the remark this answers is not in the document]
Comment (dtebbs) [reply src/snarks/groth16/mpc_utils.cpp]: Ah, yes.  There are some of those still remaining.  I'll try to do a grep and remove them all.

File: src/snarks/groth16/evaluation_from_lagrange.hpp [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+#include <map>
+
+/// Given a sequence of powers of x (with some factor) encoded in
+/// GroupT, compute the values of various linear combination of
+/// Lagrange polynomials at x.
+///
+/// TODO: Here, we assume that there may be more Lagrange polynomails
+/// than variables in the original qap, so we use a map to avoid
+/// evaluating those that are unused.  This may not be the optimal approach.
+template<typename ppT, typename GroupT> class evaluation_from_lagrange
```
Comment (AntoineRondelet) [inline src/snarks/groth16/evaluation_from_lagrange.hpp]: I'd rename this class to something like `evaluator_from_lagrange` to better refer to the notion of "an object". Then, this `evaluator_from_lagrange` could `evaluate_from_langrange_factors`

File: src/snarks/groth16/mpc_utils.hpp [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+
+/// Output from the second step of the first layer - the linear
+/// combination of elements in powersoftau, based on a specific
+/// circuit.
+class mpc_layer1
```
Comment (AntoineRondelet) [inline src/snarks/groth16/mpc_utils.hpp]: What about renaming to `mpc_layer_result` maybe?

File: src/snarks/groth16/powersoftau_utils.hpp [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+
+namespace libzeth
+{
+
+using ppT = libff::default_ec_pp;
+
+/// Output from the first phase of the MPC (powersoftau).  The
+/// structure matches that data exactly (no indication of degree,
+/// etc), so that it can be loaded directly.
+class powersoftau
```
Comment (AntoineRondelet) [inline src/snarks/groth16/powersoftau_utils.hpp]: What about renaming this class as `powersoftau_srs(/result/output)`?

File: src/snarks/groth16/evaluation_from_lagrange.hpp [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+{
+public:
+    evaluation_from_lagrange(
+        const std::vector<GroupT> &powers,
+        libfqfft::evaluation_domain<libff::Fr<ppT>> &domain);
+
+    GroupT evaluate_from_langrange_factors(
```
Comment (AntoineRondelet) [inline src/snarks/groth16/evaluation_from_lagrange.hpp]: `evaluate_from_langrange_factors` -> `evaluate_from_lagrange_factors`

File: pyClient/testZethUtils.py (hunk 5 above)
Comment (AntoineRondelet) [reply pyClient/testZethUtils.py]: Otherwise we can forget about the branch `mpc` which does not seem to diverge from develop, and this branch will become the target branch of the other mpc related PRs, and will then be the last one to be merged in develop once everything related to the MPC has been incorporated here.

Comment (dtebbs) [reply pyClient/testZethUtils.py]: Yeah, I hit the same thing.  The mpc branch should already be level with develop, but for some reason these diffs did not update.  I hoped it would update after some time delay, but I guess not.  Perhaps when I next push to this branch it will cull those extra commits..

File: src/snarks/groth16/mpc_utils.cpp (hunk 6 above)
Comment (dtebbs) [reply src/snarks/groth16/mpc_utils.cpp]: This holds only the linear combination data.  It does need a better name.

The paper indexes the layers $C_i$ and $L_i$ and uses phrases such as "We fix a layer $l$ and denote $C=C_l$ and $L=L_l$ ...".  So this is the output from the "second layer", or $L_1$, or the "Linear combinations between phases".

How about `mpc_layer_L1` or `mpc_linear_combination`?  Let me know what you think would be a good name.

File: CMakeLists.txt [hunk 15]
```diff
@@ -92,6 +92,21 @@ option(
...
+
+option(
+  USE_PT_COMPRESSION
+  "Use point compression"
+  OFF
+)
+
+set(USE_PT_COMPRESSION OFF)
+
+
```
[the remark this answers is not in the document]
Comment (dtebbs) [reply CMakeLists.txt]: Removed this line.

File: src/snarks/groth16/evaluation_from_lagrange.hpp [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+        const std::map<size_t, libff::Fr<ppT>> lagrange_factors);
+
+private:
```
[the remark this answers is not in the document]
Comment (dtebbs) [reply src/snarks/groth16/evaluation_from_lagrange.hpp]: Addressed in latest.

File: src/snarks/groth16/mpc_utils.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (dtebbs) [reply src/snarks/groth16/mpc_utils.cpp]: I think I've now addressed all cases introduced by this PR.  Note, I've intentionally left a few places where the comment does not relate to the line immediately following, e.g. TODOs etc, but if you spot any extra lines then let me know.

File: src/snarks/groth16/evaluation_from_lagrange.hpp [hunk 17]
```diff
@@ -0,0 +1,29 @@
...
+
+#include "include_libsnark.hpp"
+
+#include <map>
+
+/// Given a sequence of powers of x (with some factor) encoded in
+/// GroupT, compute the values of various linear combination of
+/// Lagrange polynomials at x.
+///
+/// TODO: Here, we assume that there may be more Lagrange polynomails
```
[the remark this answers is not in the document]
Comment (dtebbs) [reply src/snarks/groth16/evaluation_from_lagrange.hpp]: Typo should be fixed now.

File: src/snarks/groth16/mpc_utils.cpp (hunk 7 above)
Comment (dtebbs) [reply src/snarks/groth16/mpc_utils.cpp]: Updated this comment.  Hopefully it's clearer now.

File: pyClient/testZethUtils.py (hunk 5 above)
Comment (dtebbs) [reply pyClient/testZethUtils.py]: Those extra commits should no longer appear in the diff.

File: src/snarks/groth16/evaluation_from_lagrange.hpp (hunk 11 above)
Comment (dtebbs) [reply src/snarks/groth16/evaluation_from_lagrange.hpp]: Renamed as suggested.

File: src/snarks/groth16/evaluation_from_lagrange.hpp (hunk 14 above)
Comment (dtebbs) [reply src/snarks/groth16/evaluation_from_lagrange.hpp]: Typo should be fixed in latest.  Thanks.

File: src/snarks/groth16/multi_exp.hpp [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+
+#include "include_libsnark.hpp"
+
+///
```
[the remark this answers is not in the document]
Comment (dtebbs) [reply src/snarks/groth16/multi_exp.hpp]: These have now been removed.

File: src/snarks/groth16/mpc_utils.hpp [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+{
+public:
+    // TOOD: Some of these can be sparse
```
[the remark this answers is not in the document]
Comment (dtebbs) [reply src/snarks/groth16/mpc_utils.hpp]: I've removed this, as it's covered by another TODO.

File: src/snarks/groth16/mpc_utils.cpp [hunk 20]
```diff
@@ -0,0 +1,215 @@
+#include "mpc_utils.hpp"
+
+#include "evaluation_from_lagrange.hpp"
+#include "multi_exp.hpp"
+
+#include <libff/algebra/scalar_multiplication/multiexp.hpp>
+
+using namespace libsnark;
```
[the remark this answers is not in the document]
Comment (dtebbs) [reply src/snarks/groth16/mpc_utils.cpp]: Remove the `using namespace`

File: src/snarks/groth16/powersoftau_utils.cpp [hunk 21]
```diff
@@ -0,0 +1,355 @@
...
+
+    switch (marker) {
+    case 0x00:
+        // zero
+        out = libff::G1<ppT>::zero();
+        break;
+
+    case 0x04: {
+        // Uncompressed
+
```
[the remark this answers is not in the document]
Comment (dtebbs) [reply src/snarks/groth16/powersoftau_utils.cpp]: Removed the lines between case clauses.

File: src/test/powersoftau_test.cpp [hunk 22]
```diff
@@ -0,0 +1,273 @@
...
+    ASSERT_EQ(g2_1.Z, g2_1_read.Z);
+    ASSERT_EQ(G2::one(), g2_1_read);
+    ASSERT_EQ(g2_7_inv, g2_7_inv_read);
+}
+
+TEST(PowersOfTauTests, ReadPowersOfTauOutput)
+{
+    char *zeth = getenv("ZETH");
+    const std::string filename = std::string(zeth == nullptr ? "." : zeth) +
+                                 "/testdata/powersoftau_response.4.bin";
```
[the remark this answers is not in the document]
Comment (dtebbs) [reply src/test/powersoftau_test.cpp]: Yes, exactly.  This is what I used to test the code that loads the powersoftau-formatted data, and later it's used to test the `powersoftau` command.

File: src/snarks/groth16/powersoftau_utils.cpp [hunk 23]
```diff
@@ -0,0 +1,355 @@
...
+{
+    uint8_t marker;
+    in.read((char *)&marker, 1);
+
+    switch (marker) {
+    case 0x00:
+        // zero
+        out = libff::G1<ppT>::zero();
+        break;
+
```
[the remark this answers is not in the document]
Comment (dtebbs) [reply src/snarks/groth16/powersoftau_utils.cpp]: The tool seems not to have a preference, so I've removed the lines.

File: src/snarks/groth16/mpc_utils.cpp (hunk 6 above)
Comment (AntoineRondelet) [reply src/snarks/groth16/mpc_utils.cpp]: I see, thanks for clarifying.
I think `mpc_linear_combination` relates well to the "Linear combinations between phases", so this one looks good.

Comment (AntoineRondelet) [reply src/snarks/groth16/mpc_utils.cpp]: I see thanks for clarifying!
Maybe `srs_mpc_layer_L1` (or output_mpc_layer_L1) then?

Comment (dtebbs) [reply src/snarks/groth16/mpc_utils.cpp]: OK.  I'll go for `srs_mpc_layer_L1` then.  Thanks.

File: src/snarks/groth16/mpc_utils.hpp (hunk 8 above)
Comment (dtebbs) [reply src/snarks/groth16/mpc_utils.hpp]: Templatized the mpc code, removing this.

File: src/snarks/groth16/mpc_utils.cpp (hunk 9 above)
Comment (dtebbs) [reply src/snarks/groth16/mpc_utils.cpp]: Done in latest.

File: src/snarks/groth16/mpc_utils.hpp (hunk 12 above)
Comment (dtebbs) [reply src/snarks/groth16/mpc_utils.hpp]: Reamed to `srs_mpc_layer_L1` as discussed.

File: src/snarks/groth16/evaluator_from_lagrange.tcc [hunk 24]
```diff
@@ -0,0 +1,41 @@
...
+namespace libzeth
+{
+
+template<typename ppT, typename GroupT>
+evaluator_from_lagrange<ppT, GroupT>::evaluator_from_lagrange(
+    const std::vector<GroupT> &powers,
+    libfqfft::evaluation_domain<libff::Fr<ppT>> &domain)
+    : powers(powers), domain(domain)
+{
+    assert(powers.size() >= domain.m - 1);
```
Comment (AntoineRondelet) [inline src/snarks/groth16/evaluator_from_lagrange.tcc]: Shouldn't it be a strict inequality here?
The lagrange polynomials with regard to the domain S have a degree bound by `|S| - 1`. Hence, they have at most |S| terms, and each of this term represents a power of `x`, `x^i` for `i \in {0, ..., |S| - 1}`. Maybe I am missing something @dtebbs ?

File: src/snarks/groth16/powersoftau_utils.hpp (hunk 13 above)
Comment (dtebbs) [reply src/snarks/groth16/powersoftau_utils.hpp]: Renamed as suggested.

File: src/snarks/groth16/evaluator_from_lagrange.tcc (hunk 24 above)
Comment (dtebbs) [reply src/snarks/groth16/evaluator_from_lagrange.tcc]: Yes.  Fixed and added a comment.  Thanks.

File: src/test/mpc_test.cpp [hunk 25]
```diff
@@ -0,0 +1,249 @@
...
+                             alpha * qap_evaluation.Bt[i] +
+                             qap_evaluation.Ct[i];
+            ASSERT_EQ(ABC_i * G1::one(), layer1.ABC_g1[i]);
+        }
+    }
+}
+
+TEST(MPCTests, Layer2)
+{
+    const size_t n = 16;
```
Comment (AntoineRondelet) [inline src/test/mpc_test.cpp]: Why define `n` and not pass `qap.degree()` as argument to `dummy_powersoftau_from_secrets`?
Is it to get a power of 2 that is higher or eq to the circuit size in order to force the selection of a `basic_radix_2` domain?

If so, I'd probably add a comment for that, and even (maybe?) an assert `assert(qap.degree() <= n)` to make it clearer to the reader and avoid annoyance if we decided to modif the `simple_test` for any reason in the future.

File: src/snarks/groth16/mpc_utils.hpp [hunk 26]
```diff
@@ -0,0 +1,79 @@
...
+
+namespace libzeth
+{
+
+/// Output from linear combination $L_1$ - the linear combination of
+/// elements in powersoftau, based on a specific circuit.
+template<typename ppT> class srs_mpc_layer_L1
+{
+public:
+    /// { [ t(x) . x^i ]_1 }  i = 0 .. n-1
```
Comment (AntoineRondelet) [inline src/snarks/groth16/mpc_utils.hpp]: `/// { [ t(x) . x^i ]_1 }  i = 0 .. n-1` -> It should be `i = 0 .. n-2` and not i = 0 .. n-1, right?

File: src/snarks/groth16/mpc_utils.tcc [hunk 27]
```diff
@@ -0,0 +1,208 @@
...
+    libsnark::r1cs_constraint_system<libff::Fr<ppT>> &&cs,
+    const libsnark::qap_instance<libff::Fr<ppT>> &qap)
+{
+    using Fr = libff::Fr<ppT>;
+    using G1 = libff::G1<ppT>;
+    using G2 = libff::G2<ppT>;
+
+    const Fr delta_inverse = delta.inverse();
+
+    // { H_i } = { [ t(x) . x^i / delta ]_i } i = 0 .. m-1
```
Comment (AntoineRondelet) [inline src/snarks/groth16/mpc_utils.tcc]: Shouldn't it be `i = 0 .. n-2` ?

File: src/snarks/groth16/mpc_utils.tcc [hunk 28]
```diff
@@ -0,0 +1,208 @@
...
+    for (size_t i = 0; i < num_orig_ABC; ++i) {
+        B_i[i] = libsnark::knowledge_commitment<G2, G1>(
+            layer1.B_g2[i], layer1.B_g1[i]);
+    }
+
+    assert(B_i.size() == qap.num_variables() + 1);
+
+    // { L_i } = [ { ABC_i / delta } ]_1, i = l+1 .. num_variables
+    libff::G1_vector<ppT> L_g1(num_L_elements);
+    for (size_t i = 0; i < num_L_elements; ++i) {
```
Comment (AntoineRondelet) [inline src/snarks/groth16/mpc_utils.tcc]: May be worth changing the bounds of the iterations for clarification:
`for (size_t i = num_inputs + 1; i < num_variables + 1; ++i) {`. This would avoid to "manually" shift by `num_inputs` the indexes in the loop (and would be closer to the paper, which makes it easier to review)

File: src/snarks/groth16/mpc_utils.tcc [hunk 29]
```diff
@@ -0,0 +1,208 @@
...
+    assert(pot.tau_powers_g1.size() >= 2 * n - 1);
+
+    // n+1 coefficients of t
+    std::vector<Fr> t_coefficients(n + 1, Fr::zero());
+    qap.domain->add_poly_Z(Fr::one(), t_coefficients);
+
+    // Compute [ t(x) . x^i ]_1 for i = 0 .. n-2
+    libff::G1_vector<ppT> t_x_pow_i(n - 1);
+    for (size_t i = 0; i < n - 1; ++i) {
+        // Use { [x^i] , ... , [x^(i+order_L+1)] } with coefficients
```
Comment (AntoineRondelet) [inline src/snarks/groth16/mpc_utils.tcc]: `order_L` is not defined

File: src/test/mpc_test.cpp (hunk 25 above)
Comment (dtebbs) [reply src/test/mpc_test.cpp]: Yeah, good point.  Added the comment and assert.

File: src/snarks/groth16/mpc_utils.hpp (hunk 26 above)
Comment (dtebbs) [reply src/snarks/groth16/mpc_utils.hpp]: Yes.  Changed to n-2.

File: src/snarks/groth16/mpc_utils.tcc (hunk 27 above)
Comment (dtebbs) [reply src/snarks/groth16/mpc_utils.tcc]: Yes.  Changed to n-2.

File: src/snarks/groth16/mpc_utils.tcc (hunk 28 above)
Comment (dtebbs) [reply src/snarks/groth16/mpc_utils.tcc]: Changed.  (note, the indices must be similarly shifted on the other side now, but if this is more natural then let's go with that).

File: src/snarks/groth16/mpc_utils.tcc (hunk 29 above)
Comment (dtebbs) [reply src/snarks/groth16/mpc_utils.tcc]: Replaced.

**Comment by dtebbs** [comment]
Thanks for all the comments and the PR @AntoineRondelet.  I think all remaining points have been addressed now.

File: src/snarks/groth16/mpc_utils.tcc (hunk 28 above)
Comment (AntoineRondelet) [reply src/snarks/groth16/mpc_utils.tcc]: That's right indeed. I think it's clearer to see that we iterate from `l+1` to `m` in the `for` though, as this directly echoes the bounds of the set of indexes in the paper.

File: src/snarks/groth16/mpc_utils.tcc [hunk 30]
```diff
@@ -0,0 +1,212 @@
...
+    for (size_t i = 0; i < num_orig_ABC; ++i) {
+        B_i[i] = libsnark::knowledge_commitment<G2, G1>(
+            layer1.B_g2[i], layer1.B_g1[i]);
+    }
+
+    assert(B_i.size() == qap.num_variables() + 1);
+
+    // { L_i } = [ { ABC_i / delta } ]_1, i = l+1 .. num_variables
+    libff::G1_vector<ppT> L_g1(num_L_elements);
+    for (size_t i = num_inputs + 1 ; i < num_variables + 1; ++i) {
```
Comment (AntoineRondelet) [inline src/snarks/groth16/mpc_utils.tcc]: I think the build is complaining because of the extra space before `;` in this for:
`for (size_t i = num_inputs + 1 ; i < num_variables + 1; ++i) {` -> for (size_t i = num_inputs + 1; i < num_variables + 1; ++i) {

Comment (AntoineRondelet) [reply src/snarks/groth16/mpc_utils.tcc]: Fixed in next commit

Comment (dtebbs) [reply src/snarks/groth16/mpc_utils.tcc]: Thanks for fixing that.
